lib/pbench/server/database/models/datasets.py

```diff
@@ -857,21 +846,44 @@ class Metadata(Database.Base):
     # {"dashboard.seen": True}
     DASHBOARD = "dashboard.*"
 
-    # USER is arbitrary data saved on behalf of the owning user, as a JSON
-    # document.
+    # USER_FAVORITE allows direct access (e.g., in /datasets/list) to the
+    # user-specific "favorite" metadata. NOTE: for now, this is hardcoded as
+    # False until we have a framework for user-specific metadata.
     #
-    # Note that the hierarchical key management in the getvalue and setvalue
-    # static methods (used consistently by the API layer) allow client code
-    # to interact with these either as a complete JSON document ("user") or
-    # as a full dotted key path ("user.contact.name.first") or at any JSON
-    # layer between.
+    # {"user.favorite": False}
+    USER_FAVORITE = "user.favorite"
+
+    # USER_ACCESS allows direct access (e.g., in /datasets/list) to the set of
+    # access policies the authenticated user is allowed to set. Initially this
+    # is always just private and public; in the future this can be extended to
+    # support groups.
     #
-    # API keyword validation uses the trailing ".*" here to indicate that only
-    # the first element of the path should be validated, allowing the client
-    # a completely uninterpreted namespace below that.
+    # {"user.access": ["private", "public"]}
+    USER_ACCESS = "user.access"
+
+    # USER_THRESHOLD allows direct access (e.g., in /datasets/list) to the user
+    # preference setting controlling advance notice of expiring datasets in the
+    # Overview page. This will be mapped to user profile metadata when that's
+    # available.
```

Comment:
Once again, I'm not understanding how this is supposed to work.  That is, I can imagine the user (or the Dashboard on the user's behalf) wanting to list all of the datasets which are going to expire within the user's default threshold.  To do that, the dashboard would do a `/datasets/list` query with an "expiring before" qualifier.  To come up with a value for that qualifier, the dashboard would request the user setting for the threshold.  I can see how this value might be stored under a `user.threshold` key, but I don't see any particular reason to architect that.  The dashboard is likely to have a number of pieces of context that it wants to store and retrieve on behalf of the user...why single this one out?

If the server needs special access to it, then that would be a valid reason, but I would advise against that level of coupling.  (I don't think it is overburdensome to expect that the Dashboard would handle this.)

Reply:
Again, valid points, and putting it here just seemed like a nice convenience; and an expedience since we have the mechanism, and they're calling it, and we have no user profile to deal with this.

We need some server support to persist user preference settings like this; and while it's possible they'll only be used at the dashboard, I think there's probably a use case for something like an email to warn about expiring datasets rather than count on users logging in. So a server supported user preference seemed "preferable" to having the dashboard chuck it somewhere under the "dashboard" key namespace like "saved" and "seen".

However, like the IAM access values, this isn't something that differs between datasets, and in fact isn't directly related to any of them. Dang. So it should be either part of the user profile, or part of an extended user metadata model which we don't have.

And while not as excessive as returning the access list for every dataset, it's still a bit inefficient.

Reply:
So, I'm thinking about how this value would be used.

What I _think_ we want to do is have the server do a single, complex database query to return all of the datasets which are "about to expire".  If each dataset had an expiration date and a threshold, then perhaps we could make a query for all entries whose expiration minus its threshold is before today.  In order for this to be possible, the threshold has to be an attribute of the dataset, stored with the dataset.

What I'm pretty sure that we don't want to do is to iterate over each dataset, locate its user, look up the threshold associated with that user, and then check the expiration date.  That sounds crushingly slow.

It would be slightly better to iterate over each user, grab its threshold, add it to the current date, and then do a database lookup for all datasets owned by that user with expirations less than the target date.  That's still not as nice as doing it in a single (albeit, complex) query.  But, this would allow us to associate the threshold with the user instead of with the dataset.

Nevertheless, I think I would recommend having the threshold be per-dataset attribute which is set (on upload) from a default which is user attribute (unless it is specified explicitly).  (The user can update the dataset's threshold later with a dataset-metadata change.)  I think this will ultimately be more flexible and easier to administer, even if it does take up more space (presuming that there are many more datasets than there are users).

In any case, I don't think the threshold value belongs in the `user` namespace.  There should probably be one in the `dataset` namespace for each dataset, and a default value (if it's settable on a per-user basis) should be stored in the user profile along with the user's email address.

Reply:
> My intent with `datasets/list` was that it be metadata-driven, and in theory you should be able to select based on whatever metadata you can show.

That sounds great, but "metadata" in this context probably has to be limited to _dataset_ metadata; user/dashboard/client and server metadata wouldn't be applicable, right?

> But we still need a persisted user preference for the threshold. That could be `dashboard.threshold`, completely under the dashboard's control; but if so I wouldn't want the server to understand it.

I concur.

> If we want to entertain the option of, for example, sending warning emails, (and I kinda like that idea), I'd prefer to have it in an architected user metadata; whether as a first-class User property or as a UserMetadata key analogous to my dataset metadata.

I'm with you on the warning emails.  So, either the threshold needs to be (per-)dataset metadata or I think we take on some serious labor (unless the value is the same for all users and/or all datasets).

I can easily imagine a desire for having different datasets to have different expiration dates (and to have expiration dates which are independent of their creation/upload dates).  I'm having more trouble imagining that different datasets (owned by the same user) would need different warning thresholds.  To be honest, having different thresholds for different users seems like a capability that we probably don't need.  I think having a single threshold set in the server configuration for all users (perhaps with multiple warnings at "soon", "very soon", and "now") would be perfectly adequate.